File: javascript/json-transform/src/functions/TransformerFunctionGroup.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { getAsString, isNullOrUndefined } from "../JsonHelpers";
import CompareBy from "./common/CompareBy";

class TransformerFunctionGroup extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        by: { type: ArgType.Transformer, position: 0 },
        order: { type: ArgType.Enum, position: 1, defaultEnum: "ASC" },
        type: { type: ArgType.Enum, position: 2, defaultEnum: "AUTO" },
        then: { type: ArgType.Array, position: 3, defaultIsNull: true },
      },
    });
  }

  add(root: Record<string, any>, by: CompareBy) {
    let elem = root;
    const byby = by.by;
    const bybySize = byby.length ?? 0;
    for (let i = 0; i < bybySize - 1; i++) {
      const byKey = byby[i];
      // when adding a grouping key, fallback on empty string if null
      const key = getAsString(byKey) ?? "";
      if (isNullOrUndefined(root[key])) {
        const jo: any = {};
        root[key] = jo;
        elem = jo;
      } else {
        elem = root[key];
      }
    }
    const byKey = byby[bybySize - 1];
    // when adding a grouping key, fallback on empty string if null
    const key = getAsString(byKey) ?? "";
    const jArr = elem[key] ?? [];
    jArr.push(by.value);
    elem[key] = jArr;
    return root;
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElementStreamer(null);
    if (value == null) {
      return {};
    }
    const type = await context.getEnum("type");
    const order = await context.getEnum("order");
    const by = await context.getJsonElement("by", false);
    if (isNullOrUndefined(by)) {
      return {};
    }

    let comparator = CompareBy.createByComparator(0, type);
    if ("DESC" === order?.toUpperCase()) {
      comparator = comparator.reversed();
    }

    const chain: any[] = [by];
    const thenArr = context.has("then") ? await context.getJsonArray("then", false) : null;
    if (thenArr != null) {
      const thenArrSize = thenArr.length;
      for (let i = 0; i < thenArrSize; i++) {
        const thenObj = thenArr[i];
        const thenType = thenObj.type ? getAsString(thenObj.type)?.trim() ?? null : null;
        const thenOrder = thenObj.order;
        let thenComparator = CompareBy.createByComparator(i + 1, thenType);
        const thenDescending = !isNullOrUndefined(thenOrder) && getAsString(thenOrder)?.toUpperCase() === "DESC";
        if (thenDescending) {
          thenComparator = thenComparator.reversed();
        }
        comparator = comparator.then(thenComparator.compare);
        chain.push(thenObj.by);
      }
    }

    const result: any = {};
    await value
      .stream()
      .map(async item => {
        const cb = new CompareBy(item);
        for (const jsonElement of chain) {
          const byKey = await context.transformItem(jsonElement, item);
          cb.by.push(byKey);
        }
        return cb;
      })
      .sortedWith(comparator.compare)
      .forEach(itm => this.add(result, itm));
    return result;
  }
}

export default TransformerFunctionGroup;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionGroup.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.adapters.JsonAdapter;
import co.nlighten.jsontransform.functions.common.*;

import java.util.ArrayList;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicReference;

public class TransformerFunctionGroup extends TransformerFunction {

    public TransformerFunctionGroup() {
        super(FunctionDescription.of(
            Map.of(
            "by", ArgumentType.of(ArgType.Transformer).position(0),
            "order", ArgumentType.of(ArgType.Enum).position(1).defaultEnum("ASC"),
            "type", ArgumentType.of(ArgType.Enum).position(2).defaultEnum("AUTO"),
            "then", ArgumentType.of(ArgType.Array).position(3).defaultIsNull(true)
            )
        ));
    }

    private Object add(JsonAdapter<?, ?, ?> adapter, Object root, CompareBy by) {
        var elem = new AtomicReference<>(root);
        var byby = by.by;
        var bybySize = byby.size();
        for (var i = 0; i < bybySize - 1; i++) {
            var byKey = byby.get(i);
            // when adding a grouping key, fallback on empty string if null
            var key = adapter.isNull(byKey) ? "" : adapter.getAsString(byKey);
            var val = adapter.get(root, key);
            elem.set(Objects.requireNonNullElseGet(adapter.isJsonObject(val) ? val : null, () -> {
                var jo = adapter.createObject();
                adapter.add(elem.get(), key, jo);
                return jo;
            }));
        }
        var byKey = byby.get(bybySize - 1);
        // when adding a grouping key, fallback on empty string if null
        var key = adapter.isNull(byKey) ? "" : adapter.getAsString(byKey);
        var jArr = Objects.requireNonNullElseGet(
                adapter.get(elem.get(), key),
                adapter::createArray);
        adapter.add(jArr, by.value);
        adapter.add(elem.get(), key, jArr);
        return root;
    }

    @Override
    public CompletionStage<Object> apply(FunctionContext context) {
        var adapter = context.getAdapter();
        var value = context.getJsonElementStreamer(null);
        if (value == null) {
            return adapter.createObject();
        }
        var type = context.getEnum("type");
        var order = context.getEnum("order");
        var by = context.getJsonElement("by", false);
        if (adapter.isNull(by)) {
            return adapter.createObject();
        }

        var chain = new ArrayList<>();

        var comparator = CompareBy.createByComparator(adapter, 0, type);
        if ("DESC".equalsIgnoreCase(order)) {
            comparator = comparator.reversed();
        }
        chain.add(by);

        var thenArr = context.has("then") ? context.getJsonArray("then", false) : null;
        if (thenArr != null) {
            var thenArrSize = adapter.size(thenArr);
            for (var i = 0; i < thenArrSize; i++) {
                var thenObj = adapter.get(thenArr, i);
                var thenType = adapter.has(thenObj, "type") ? context.getAsString(adapter.get(thenObj, "type")).trim() : null;
                var thenOrder = adapter.get(thenObj,"order");
                var thenComparator = CompareBy.createByComparator(adapter,i + 1, thenType);
                var thenDescending = !adapter.isNull(thenOrder) && context.getAsString(thenOrder).equalsIgnoreCase("DESC");
                if (thenDescending) {
                    thenComparator = thenComparator.reversed();
                }
                comparator = comparator.thenComparing(thenComparator);
                chain.add(adapter.get(thenObj,"by"));
            }
        }

        var result = adapter.createObject();
        value.stream()
                .map(item -> {
                    var cb = new CompareBy(item);
                    cb.by = new ArrayList<>();
                    for (var jsonElement : chain) {
                        var byKey = context.transformItem(jsonElement, item);
                        cb.by.add(byKey == null ? adapter.jsonNull() : byKey);
                    }
                    return cb;
                })
                .sorted(comparator)
                .forEachOrdered(itm -> add(adapter, result, itm));
        return result;
    }
}
